torch/profiler/profiler.py

```diff
@@ -3,12 +3,185 @@
 import os
 import tempfile
 from enum import Enum
-from typing import Any, Callable, Iterable, Optional
+from functools import partial
+from typing import Any, Callable, Dict, Iterable, List, Optional, Tuple
 from warnings import warn
 
 import torch
 import torch.autograd.profiler as prof
-from torch.autograd import kineto_available, ProfilerActivity
+from torch.autograd import ProfilerActivity, kineto_available
+
+
+def supported_activities():
+    """
+    Returns a set of supported profiler tracing activities.
+
+    Note: profiler uses CUPTI library to trace on-device CUDA kernels.
+    In case when CUDA is enabled but CUPTI is not available, passing
+    ``ProfilerActivity.CUDA`` to profiler results in using the legacy CUDA
+    profiling code (same as in the legacy ``torch.autograd.profiler``).
+    This, in turn, results in including CUDA time in the profiler table output,
+    but not in the JSON trace.
+    """
+    return torch.autograd._supported_activities()
+
+
+class kineto_profile(object):
+    """Low-level profiler wrap the autograd profile
+
+    Args:
+        activities (iterable): list of activity groups (CPU, CUDA) to use in profiling, supported values:
+            ``torch.profiler.ProfilerActivity.CPU``, ``torch.profiler.ProfilerActivity.CUDA``.
+            Default value: ProfilerActivity.CPU and (when available) ProfilerActivity.CUDA.
+        record_shapes (bool): save information about operator's input shapes.
+        profile_memory (bool): track tensor memory allocation/deallocation.
+        with_stack (bool): record source information (file and line number) for the ops.
+        with_flops (bool): use formula to estimate the FLOPS of specific operators
+            (matrix multiplication and 2D convolution).
+        with_modules (bool): record module hierarchy (including function names)
+            corresponding to the callstack of the op. e.g. If module A's forward call's
+            module B's forward which contains an aten::add op,
+            then aten::add's module hierarchy is A.B
+            Note that this support exist, at the moment, only for TorchScript models
+            and not eager mode models.
+
+    .. note::
+        Enabling shape and stack tracing results in additional overhead.
+        When record_shapes=True is specified, profiler will temporarily hold references to the tensors;
+        that may further prevent certain optimizations that depend on the reference count and introduce
+        extra tensor copies.
+    """
+    def __init__(
+            self,
+            *,
+            activities: Optional[Iterable[ProfilerActivity]] = None,
+            record_shapes: bool = False,
+            profile_memory: bool = False,
+            with_stack: bool = False,
+            with_flops: bool = False,
+            with_modules: bool = False):
+        if activities:
```

Comment:
This can just be `activities = set(activities) or supported_activities()`

(I realize that this is just moving the old code around, but might as well do some housecleaning in the process.)

Reply:
The mypy check would failure if we do the assignment since the type hinting for activities is `Optional[Iterable[ProfilerActivity]]`. The type hinting of `set(activities)` or `supported_activities()` is `Set[ProfilerActivity]`.

Reply:
will do

Reply:
Actually, I was wrong. It has to be `activities = set(activities or supported_activities())` because `set(None)` is not allowed. (I feel like I saw a comment where someone else noticed that, but I can't find it.)

Reply:
the current logic is: `self.activities = set(activities) if activities else supported_activities()` . Need we change it to `self.activities = set(activities or supported_activities())`?

Reply:
Gah! I completely missed that distinction. Ok, I'll start importing it. Thanks for your patience.